src/asound/bluealsa-ctl.c

```diff
@@ -992,23 +1120,90 @@ SND_CTL_PLUGIN_DEFINE_FUNC(bluealsa) {
 		goto fail;
 	}
 
-	if (!bluealsa_dbus_get_pcms(&ctl->dbus_ctx, &ctl->pcm_list, &ctl->pcm_list_size, &err)) {
+	if (!bluealsa_dbus_get_pcms(&ctl->dbus_ctx, &pcm_list, &pcm_list_size, &err)) {
 		SNDERR("Couldn't get BlueALSA PCM list: %s", err.message);
 		ret = -ENODEV;
 		goto fail;
 	}
 
+	if (bacmp(&device, BDADDR_ALL) != 0) {
+		struct ba_pcm *_pcms = NULL;
+		size_t n, count = 0;
+
+		ctl->single_device = true;
+
+		if (bacmp(&device, BDADDR_ANY) == 0) {
+			/* Interpret BDADDR_ANY as a request for the most
+			 * recently connected bluetooth audio device */
+			uint32_t seq = 0;
+			size_t latest = 0;
+
+			if (pcm_list_size == 0) {
+				SNDERR("No BlueALSA audio devices connected");
+				ret = -ENODEV;
+				goto fail;
+			}
+
+			for (n = 0; n < pcm_list_size; n++) {
+				if (pcm_list[n].sequence >= seq) {
+					seq = pcm_list[n].sequence;
+					latest = n;
+				}
+			}
+
+			bacpy(&device, &pcm_list[latest].addr);
+		}
+
+		for (n = 0; n < pcm_list_size; n++) {
+			if (bacmp(&device, &pcm_list[n].addr) == 0) {
+				struct ba_pcm *tmp = _pcms;
+				if ((tmp = realloc(tmp, (n + 1) * sizeof(*tmp))) == NULL) {
+					SNDERR("Couldn't create PCM info: %s", strerror(ENOMEM));
+					ret = -ENOMEM;
+					free(pcm_list);
+					goto fail;
+				}
+				_pcms = tmp;
+				memcpy(&_pcms[count++], &pcm_list[n], sizeof(struct ba_pcm));
+			}
+		}
+		free(pcm_list);
+		ctl->pcm_list = _pcms;
+		ctl->pcm_list_size = count;
+	}
+	else {
+		ctl->pcm_list = pcm_list;
+		ctl->pcm_list_size = pcm_list_size;
+	}
+
 	if (bluealsa_create_elem_list(ctl) == -1) {
 		SNDERR("Couldn't create control elements: %s", strerror(errno));
 		ret = -errno;
 		goto fail;
 	}
 
+	if (ctl->single_device && pipe2(ctl->pipefd, O_CLOEXEC) == -1) {
+		SNDERR("Couldn't create event pipe: %s", strerror(errno));
+		ret = -errno;
+		goto fail;
+	}
+
 	ctl->ext.version = SND_CTL_EXT_VERSION;
 	ctl->ext.card_idx = 0;
 	strncpy(ctl->ext.id, "bluealsa", sizeof(ctl->ext.id) - 1);
 	strncpy(ctl->ext.driver, "BlueALSA", sizeof(ctl->ext.driver) - 1);
-	strncpy(ctl->ext.name, "BlueALSA", sizeof(ctl->ext.name) - 1);
+
+	if (ctl->single_device) {
+		if (ctl->dev_list_size != 1) {
+			SNDERR("No such audio device: %s", device_str);
+			ret = -ENODEV;
+			goto fail;
+		}
+		strncpy(ctl->ext.name, ctl->dev_list[0]->name, sizeof(ctl->ext.name));
```

Comment:
I think that it will be safer to assume that ALSA expect zero-terminated string in `name`. So, please use `sizeof(ctl->ext.name) - 1`.